### PR TITLE
Features/elevation provider for pgm data

### PR DESCRIPTION
**Pull request by Ciszq**

### CODE REVIEW

File: src/NSrtm.Core.xTests/PgmElevationProviderCoordinatesFindingTests.cs [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+{
+    public class PgmElevationProviderCoordinatesFindingTests
+    {
+        [Theory]
+        [MemberData("SimplePgmCellCoords")]
+        internal void ExtractingValuesFromSimpleSegmentsWorksCorrectly(
+            PgmCellCoords coords,
+            double latIncrement,
+            double lonIncrement,
+            IEnumerable<PgmCellCoords>
```
Comment (kwesolowski) [inline src/NSrtm.Core.xTests/PgmElevationProviderCoordinatesFindingTests.cs]: This formatting is ugly, please review layout

File: src/NSrtm.Core/Pgm/PgmCellCoords.cs [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+        }
+
+        public static bool operator !=(PgmCellCoords left, PgmCellCoords right)
+        {
+            return !left.Equals(right);
+        }
+
+        public static PgmCellCoords ForCoordinatesUsingDescription(double latitude, double longitude, PgmDataDescription dataDescription)
+        {
+            var nodeVerticalPosition = Math.Floor(latitude / dataDescription.LatitudeIncrementDegrees) * dataDescription.LatitudeIncrementDegrees;
```
Comment (kwesolowski) [inline src/NSrtm.Core/Pgm/PgmCellCoords.cs:52]: It is deliberate that for negative numbers you would like to rount towards -infinity, not zero?

File: src/NSrtm.Core/Pgm/PgmElevationProvider.cs [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+                                                       .ToList();
+            return BicubicCalculator.GetSpline(values: formattedUndulations, step: dataDescription.LatitudeIncrementDegrees);
+        }
+
+        private static PgmCellCoords normalizeCoords(double lat, double lon)
+        {
+            var normalizedLongitude = lon;
+            if (lat > 90 || lat < -90)
+            {
+                normalizedLongitude += 180;
```
Comment (kwesolowski) [inline src/NSrtm.Core/Pgm/PgmElevationProvider.cs]: Why `lat` value affects `normalizedLongitude?

File: src/NSrtm.Core/Pgm/PgmElevationProvider.cs [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+            }
+            normalizedLongitude = (360 + (normalizedLongitude % 360)) % 360;
```
Comment (kwesolowski) [inline src/NSrtm.Core/Pgm/PgmElevationProvider.cs]: This is simplest solution?

File: src/NSrtm.Core/Pgm/PgmElevationProvider.cs [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+            var absoluteLat = Math.Abs(Math.Abs(lat % 180) - 90);
```
Comment (kwesolowski) [inline src/NSrtm.Core/Pgm/PgmElevationProvider.cs]: again simplest one?

File: src/NSrtm.Core/Pgm/PgmElevationProvider.cs [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+        {
+            var longitudeInEgmDatum = longitude + 180;
+            var mainNode = PgmCellCoords.ForCoordinatesUsingDescription(latitude, longitudeInEgmDatum, dataDescription);
+            var interpolatedCell = _continuousSurface.GetOrAdd(mainNode, getInterpolationForCellSurface);
+            return interpolatedCell.Evaluate(latitude, longitudeInEgmDatum);
+        }
+
+        public Task<double> GetElevationAsync(double latitude, double longitude)
+        {
+            return Task.FromResult(GetElevation(latitude, longitude));
```
Comment (kwesolowski) [inline src/NSrtm.Core/Pgm/PgmElevationProvider.cs:75]: First call would block thread for a few seconds?

File: src/NSrtm.Core/Pgm/PgmElevationProvider.cs [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+        }
+
+        public Level ElevationBase { get { return dataDescription.Level; } }
+        public Level ElevationTarget { get { return Level.EllipsoidWgs84; } }
+
+        #region Static Members
+
+        internal static IEnumerable<PgmCellCoords> findCellAndSurroundingNodesCoords(PgmCellCoords pgmCellCoords, double latIncrement, double lonIncrement)
+        {
+            var horizontalNodes = Enumerable.Range(-1, 4)
```
Comment (kwesolowski) [inline src/NSrtm.Core/Pgm/PgmElevationProvider.cs]: Please replace var with explicit type.

File: src/NSrtm.Core/Pgm/PgmElevationProvider.cs [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+
+        #region Provider creation
+
+        /// <summary>
+        ///     Creates elevation provider which loads PGM file to memory and uses bicubic spline interpolation.
+        /// </summary>
+        /// <param name="path"></param>
+        /// <returns>Created provider.</returns>
+        [NotNull]
+        public static IElevationProvider CreateWithStoredGridInMemory([NotNull] string path)
```
Comment (kwesolowski) [inline src/NSrtm.Core/Pgm/PgmElevationProvider.cs]: Why `CreateWithStoredGridInMemory` and not
`CreateWithGridDataStoredInMemory`

File: src/NSrtm.Core/Pgm/PgmElevationProvider.cs [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+            return new PgmElevationProvider(pgmGeoidUndulationMemory);
+        }
+
+        /// <summary>
+        ///     Creates elevation provider which reads PGM data from file.
+        /// </summary>
+        /// <param name="path"></param>
+        /// <returns>Created provider.</returns>
+        [NotNull]
+        public static IElevationProvider CreateDirectAccessToGrid([NotNull] string path)
```
Comment (kwesolowski) [inline src/NSrtm.Core/Pgm/PgmElevationProvider.cs]: Why `CreateDirectAccessToGrid` and not
`CreateWithDirectAccessToGridData`

File: src/NSrtm.Core/Pgm/PgmCellCoords.cs (hunk 2 above)
Comment (Ciszq) [inline src/NSrtm.Core/Pgm/PgmCellCoords.cs:52]: Yes, that is way I use Floor.

File: src/NSrtm.Core/Pgm/PgmElevationProvider.cs (hunk 3 above)
Comment (Ciszq) [inline src/NSrtm.Core/Pgm/PgmElevationProvider.cs]: Becouse lat 91 lon 0 should be converted to lat 89 lon 180.

File: src/NSrtm.Core/Pgm/PgmElevationProvider.cs (hunk 5 above)
Comment (Ciszq) [inline src/NSrtm.Core/Pgm/PgmElevationProvider.cs]: Maybe you can simplify this. If you want read how to ganerate triangle wave i recommend this article http://www.hitmaroc.net/38786-5090-there-one-line-function-that-generates-triangle-wave.html

File: src/NSrtm.Core/Pgm/PgmElevationProvider.cs (hunk 6 above)
Comment (Ciszq) [inline src/NSrtm.Core/Pgm/PgmElevationProvider.cs:75]: PgmElevationProvider get grid in constructor. Get elevation only create bivariatepolygonal if necessary and take elevation.

**Pull request merged**